octospi driver for stm32 mcus

FRASTM opened the pull request:
This PR adds the support of the OCTOSPI peripheral and driver on the stm32 devices

A new Kconfig is defined to enable this new driver for accessing the NOR flash octo spi memories.
The new stm32 flash driver for octospi is similar to the quadspi.
A sample application is also adatpted to test the driver :  samples/drivers/spi_flash

In this configuration, only the second instance (octospi2) is enabled, The driver is not yet designed for multi-instance.
The stm32l562 disco kit and b_u585 disco kit are available for testing.
Note that the NOR octo-flash which is on the disco kit  does respond to the SFDP (wrong value is given). 
So the **sfdp-bfp table** is given by the DTS of the mx25lm51245 NOR octo flash controller. 

No DMA transfer yet

fixes https://github.com/zephyrproject-rtos/zephyr/issues/44690

Requires the https://github.com/zephyrproject-rtos/hal_stm32/pull/122

Signed-off-by: Francois Ramu francois.ramu@st.com

FRASTM committed:
dts/bindings: introduce a new octospi devicetree

Add the DTS binding for OCTOSPI interface for the stm32 devices
from STMicroelectronics.
This corresponds to a NOR octo SPI flash.
In this config, there is only on NOR-flash device.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: flash : introduce a new driver for the OCTOSPI

This is the stm32 octospi driver based on the exisitng quadspi
 for stm32 devices and source code from the STM32Cube.
This drivers initialized the peripheral and the NOR memory
in SPI or OctoSPI mode with STR or DTR data Transfer rates.
The NOR-flash can provide the SDFP table directly (if supported)
or through the DeviceTree.
 Limitation: no DMA transfer.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
boards: arm: stm32u5 disco kit target board has octo SPI instance

This commit enables the octo SPI peripheral to the flash-nor
on the target board b_u585i_iot02a from STMicroelectronics.
Note that JESD16 requires 9 dwords for the sdfp table.
The configuration is for OctoSPI bus through IO manager.
The NOR octoflash is MX25LM51245 or compatible.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

heinwessels reviewed (commented):
I'm sure you'll figure it out as well, or probably already did. I just thought I would help since it's in my interest that this gets merged too :)

FRASTM commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1039 @@
+		ospi_mgr_cfg.IOLowPort = HAL_OSPIM_IOPORT_2_LOW;
+		ospi_mgr_cfg.IOHighPort = HAL_OSPIM_IOPORT_2_HIGH;
+	}
+	if (HAL_OSPIM_Config(&dev_data->hospi, &ospi_mgr_cfg,
```
It appears that when using the OCTOSPI2 instance, the OSPIMux is unchanged (left disabled)
So even with ospi_mgr_cfg.Req2AckTime >0, the HAL_OSPIM_Config() returns HAL_OK

erwango replied:
@heinwessels, @FRASTM Is this point resolved ?

heinwessels replied:
Was just about to comment that. I believe the comment `/* arbitrary value */` can be more descriptive, but yes, point is resolved.

FRASTM commented:
OSPI init'd
OSPI M init'd
OSPI delay block init'd

heinwessels commented:
> During the ospi_read_sfdp, no command complete IT occurs, the ospi remains busy (bit 5 of the OCTOSPI_SR).

I think I know what the issue is because I struggled with the same thing. The current HAL driver `HAL_OSPI_Command_IT` does not support sending/receiving data. I had to change my driver to rather use `HAL_OSPI_Command`. It's caused by there's the following line in `stm32h7xx_hal_ospi.c:983`:
```
if (... && (cmd->DataMode == HAL_OSPI_DATA_NONE) && ..){
    // Do command
}
else
{
    status = HAL_ERROR;
    hospi->ErrorCode = HAL_OSPI_ERROR_INVALID_SEQUENCE;
}
```

If you read the `hospi->ErrorCode` you should see it set to `HAL_OSPI_ERROR_INVALID_SEQUENCE`.

FRASTM commented:
Seems that the HAL_OSPI_DLYB_SetConfig fails  with the units=56 and Sel =2 : after modifying the Delay, 
the Length valid flag is reset ==> "Length value in LNG is not valid."

FRASTM commented:
rebase 
add the function to have the config like the stm32Cube, reading the SFDP gives the same result as the Cube example
```
reading SFDP
SFDP magic 88888888  invalid
```

manoj153 commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1161 @@
+
+	/* Initializes the independent peripherals clock */
+//	__HAL_RCC_OSPI_CONFIG(RCC_OSPICLKSOURCE_PLL1); /* PLL1 is the clock source */
+	__HAL_RCC_OSPI_CONFIG(RCC_OSPICLKSOURCE_SYSCLK); /* */
+	__HAL_RCC_OSPI2_CLK_ENABLE();
+	__HAL_RCC_OSPI2_FORCE_RESET();
+	__HAL_RCC_OSPI2_RELEASE_RESET();
+
```
Clock related stuff is supposed to be here, or is it just here for quick testing?

FRASTM replied:
yes,
Done now by 
```
	if (clock_control_on(DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE),
			     (clock_control_subsys_t) &dev_cfg->pclken) != 0) {
		LOG_DBG("Could not enable OSPI clock");
		return -EIO;
	}
```

abelradenac commented:
Hello,
Our team is quite interested with this octospi driver proposal, and I would like to ask a few questions:
- Do you have an idea of when it could be available in the main Zephyr branch ?
- We will be using a flash memory MX25R6435F (quadspi interface). Is this octospi driver configurable for quadspi ? We are using the stm32u585 that must use the octospi interface to address mutli I/O SPI peripherals (unlike the stm32l4 series that also has a dedicated quadspi interface).
- And finally is there something we can do to help you complete this draft version ?

FRASTM committed:
module: Kconfig stm32 includes the DELAYBLOCK (DLYB)

add the config to support the DELAYBLOCK LL driver
for stm32 mcus like stm32U5x.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
In the NOR-flash init sequence:
- adding a function to read the Flash ID (0x9F) on the stm32L562e_dk the ID 0xC2 0x85 0x3A is  returned by the NOR-flash
- adding a function to read the SFDP (0x5A) : the value received from the NOR flash (0xFFFFFFFF) is 0xFFFFFFFF not correct : init fails

FRASTM committed:
dts: arm: stm32l5 as a octospi node

Add the octospi node to the stm32l5 family

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
> Would it be possible to add a `dts` property to allow to set the data width to something other than one line? Currently it's hardcoded with `.DataMode = HAL_OSPI_DATA_1_LINE`.
> 
> Here is a [PR](https://github.com/fancom/zephyr/pull/31) on a forked repo that has implemented a flash OSPI with a datawidth of 4 lines (that's worked into the existing `flash_qspi` driver). It was a fairly easy extention to the original code to add variable datawidth.

Ok in a next version, we could have one or 4 or 8 line configuration.

Also is a DTR mode or STR mode selection to be included in the DTS as a property

FRASTM commented:
> Would it be possible to add a `dts` property to allow to set the data width to something other than one line? Currently it's hardcoded with `.DataMode = HAL_OSPI_DATA_1_LINE`.
> 
> Here is a [PR](https://github.com/fancom/zephyr/pull/31) on a forked repo that has implemented a flash OSPI with a datawidth of 4 lines (that's worked into the existing `flash_qspi` driver). It was a fairly easy extention to the original code to add variable datawidth.


--> Introducing the configuration in the DTS : 
`<data-mode>`
#define STM32_OSPI_**S**PI_MODE	1U      /* 1 Cmd Line, 1 Address Line and 1 Data Line    */
#define STM32_OSPI_**O**PI_MODE	8U      /* 8 Cmd Lines, 8 Address Lines and 8 Data Lines */
`<data-rate>`
#define STM32_OSPI_**S**TR_TRANSFER 1U /* Single Transfer Rate */
#define STM32_OSPI_**D**TR_TRANSFER 2U /* Double Transfer Rate */

FRASTM committed:
drivers: flash: spi nor header includes definition for OSPI commands

With the introduction of the OSPI NOR flash controller
more octal commands and parameters are defined.
It completes the existing SPI commands

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
include: bindings: new bindings for the ospi flash controller

The new octoSPI flash controller driver
requires parameter to configure the peripheral
especially matching the data mode and the transfer rate

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
Re-formatting the PR to introduce the octospi driver for stm32 MCUs. This version is given by the STM32Cube. Refer to the [AN5050 - Octo-SPI interface on STM32 microcontrollers](https://www.st.com/content/ccc/resource/technical/document/application_note/group0/91/dd/af/52/e1/d3/48/8e/DM00407776/files/DM00407776.pdf/jcr:content/translations/en.DM00407776.pdf)

Introducing 2 possible data modes (spi-bus-width) and 2 possible transfer rate (STR/DTR) through DTS.
Currently SPI STR and OctoSPi STR mode are supported.
The driver was tested on the disco stm32l562e_dk on a micro MX25LM51245 octoSPI  NOR  flash  in SPI/STR and OctoSPI/DSTR modes with a sample like ./samples/drivers/spi_flash adapted for the ospi.
The SDFP is not given by the NOR flash : the command is not answered by the NOR of that board.

FRASTM commented:
adding the **octoSPI** mode and double data rate **DTR**
set DTS node of the mx25lm51245: ospi-nor-flash
with
```
 		spi-bus-width = <8>;
		data-rate = <2>;
```

FRASTM committed:
dts: arm: stm32u5 as a octospi node

Add the octospi nodes to the stm32u5 family

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
supports the b_u585i_iot02a board with octo NOR flash on SPI/OPI mode and STR/DTR data transfer
(no DMA transfer)

FRASTM commented:
adding the SDFP table from the DTS because some NOR octo- flash are all 0xFF,
In that case it is needed to give the 
```
		sfdp-bfp = [
			53 46 44 50 06 01 02 ff
			00 06 01 10 30 00 00 ff
			C2 00 01 04 10 01 00 ff
			84 00 01 02 C0 00 00 ff    ];
```

in the DTS of the ospi-nor-flash node.

FRASTM committed:
boards: arm: stm32l562 disco kit target board has octo SPI instance

This commit enables the octo SPI peripheral to the flash-nor
on the target board stm32l562e_dk from STMicroelectronics.
Note that JESD16 requires 9 dwords for the sdfp table.
The NOR octoflash is MX25LM51245 or compatible.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
samples: drivers flash add a test case for stm32 octospi

This add a new sample application to test and validate
the stm32 octospi driver on the NOR octo-flash present on
stm32 disco kits
There should be a special case to erase the complete flash.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
The JES216 requires 9 dwords for the sdfp table (here set to 0)
That will make the erase type table correct for use in the setup_pages_layout(). In case the erase param is wrong (from the sdfp table, the  default value is SPI_NOR_SECTOR_SIZE)

Removing the CONFIG_DEBUG to run the sample test

FRASTM commented:
reformat file to fix warnings

FRASTM commented:
describe the st,stm32-ospi.yaml and prepare for DMA

GeorgeCGV commented:
Octo-SPI support is right on time. Unfortunately, https://github.com/zephyrproject-rtos/hal_stm32/pull/122 misses STM32H7 series.

FRASTM commented:
Th

> Octo-SPI support is right on time. Unfortunately, [zephyrproject-rtos/hal_stm32#122](https://github.com/zephyrproject-rtos/hal_stm32/pull/122) misses STM32H7 series.

Support of the stm32H7 serie will come soon

erwango commented on include/zephyr/dt-bindings/flash_controller/stm32_ospi.h:
```diff
@@ -0,0 +1,36 @@
+#define STM32_OSPI_DTR_TRANSFER                 2U
+
+/* Max Time value during reset or erase operation */
+#define STM32_OSPI_RESET_MAX_TIME               100U
+#define STM32_OSPI_BULK_ERASE_MAX_TIME          460000U
+#define STM32_OSPI_SECTOR_ERASE_MAX_TIME        1000U
+#define STM32_OSPI_SUBSECTOR_4K_ERASE_MAX_TIME  400U
+#define STM32_OSPI_WRITE_REG_MAX_TIME           40U
```
It doesn't seem those are bt-bindings but rather drivers specific definitions, so they should be put in driver the specific header (or directly in the header)

FRASTM replied:
moved to drivers/flash/flash_stm352_ospi.c

erwango commented on boards/arm/stm32l562e_dk/stm32l562e_dk_common.dtsi:
```diff
@@ -127,6 +127,47 @@
+		reg = <0>;
+		ospi-max-frequency = <50000000>;
+		size = <DT_SIZE_M(4)>;  /* 512 KBytes */
+		spi-bus-width = <8>;
+		data-rate = <2>;
```
Can you use STM32_OSPI_OSPI_MODE and STM32_OSPI_STR_TRANSFER ?
Otherwise, not need to define them in a dt-binding header

FRASTM replied:
yes, 
OSPI / STR is a possible configuration
as well as OSPI/DTR and SPI/STR

FRASTM commented on boards/arm/stm32l562e_dk/stm32l562e_dk.dts:
```diff
@@ -19,6 +19,8 @@
 		zephyr,sram = &sram0;
 		zephyr,flash = &flash0;
+		zephyr,flash-controller = &mx25lm51245;
+
```
removed

FRASTM commented on drivers/flash/jesd216.h:
```diff
@@ -19,7 +19,8 @@
  */
 
 #define JESD216_CMD_READ_SFDP   0x5A
-#define JESD216_CMD_BURST_SFDP  0x5B
```
restored

FRASTM commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -64,6 +64,9 @@ struct flash_stm32_ospi_config {
 	int data_mode; /* SPI or QSPI or OSPI */
 	int data_rate; /* DTR or STR */
 	const struct pinctrl_dev_config *pcfg;
+#if DT_NODE_HAS_PROP(DT_INST(0, st_stm32_ospi_nor), sfdp_bfp)
```
with 2 commits the support of the sdfp table is not mixed with other purpose

FRASTM replied:
squashed

GeorgeCGV commented on dts/bindings/flash_controller/st,stm32-ospi-nor.yaml:
```diff
@@ -0,0 +1,57 @@
+      type: phandle-array
+      required: false
+      description: RESETn pin
+    spi-bus-width:
+      type: int
+      default: 8
+      required: false
+      description: The width of (Octo)SPI bus to which flash memory is connected.
+                   Now only value of 8 (default) or 1 (not really useful) are supported.
+      enum:
+        - 1 #  SPI mode
+        - 8 #  Octo SPI mode
```
What about `4`? Some devices have Octo-SPI and that is the only option to have Quad-SPI (_e.g. H730_).

FRASTM commented on dts/bindings/ospi/st,stm32-ospi.yaml:
```diff
@@ -0,0 +1,71 @@
+      required: false
+      description: |
+        FLash ID number. This number, if defined, helps to select the right
+        OSPI GPIO banks (defined as 'octospi_bk[12]' in pinctrl property)
```
Removed as the FLASH_ID_1 is fixed in the driver.
This <flash-id> is reserved for future use.

FRASTM commented on dts/bindings/ospi/st,stm32-ospi.yaml:
```diff
@@ -0,0 +1,71 @@
+        Optional DMA channel specifier. If DMA should be used, specifier should
+        hold a phandle reference to the dma controller, the channel number,
+        the slot number, channel configuration and finally features.
+        Even if a dmamux exists, the dma node is used (&dma1, &dma2)
```
comment is no more applicable  as there is no DMA in that driver

FRASTM commented on dts/bindings/ospi/st,stm32-ospi.yaml:
```diff
@@ -0,0 +1,71 @@
+        There is no Fifo used by this DMA peripheral.
+
+        For example dmas for TX/RX on OSPI
+           dmas = <&dma1 5 41 0x10000>;
```
comment is no more applicable as there is no DMA in that driver
But kept for next verison with DMA transfer. 
--> should the `dts/bindings/qspi/st,stm32-qspi.yaml` also be changed then, ;-) ?

erwango commented on modules/Kconfig.stm32:
```diff
@@ -549,7 +549,7 @@ config USE_STM32_LL_DAC
 	  Enable STM32Cube Digital-to-analog converter (DAC) LL module driver
 
-config USE_STM32_LL_DELAYBLOCK
+config USE_STM32_LL_DLYB
 	bool
```
Comment not addressed

FRASTM replied:
done
adding the DLYB config in addition to the DELAYBLOCK config

erwango commented on drivers/flash/Kconfig.stm32_ospi:
```diff
@@ -0,0 +1,22 @@
+# STM32 Octo SPI flash driver configuration options
+
+# Copyright (c) 2021 STMicroelctronics
```
Still there

FRASTM replied:
changed

erwango commented on drivers/flash/Kconfig.stm32_ospi:
```diff
@@ -0,0 +1,22 @@
+# STM32 Octo SPI flash driver configuration options
+
+# Copyright (c) 2021 STMicroelctronics
```
Still there

FRASTM replied:
fixed

erwango commented on dts/arm/st/u5/stm32u5.dtsi:
```diff
@@ -348,6 +348,28 @@
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "disabled";
+			label= "OSCTOPI2";
```
```suggestion
			label= "OSCTOSPI2";
```

FRASTM replied:
fixed

FRASTM committed:
drivers: flash: JESD216 ospi command for SFDP

Includes the 16bit command to read the SFDP in the NOR flash.
according to the JEDEC standard.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1787 @@
+	int cmd_status;
+#if  DT_HAS_COMPAT_STATUS_OKAY(st_stm32_dma)
+	struct stream dma;
+#endif
```
removed

FRASTM commented:
squashing commit ec4ce3b9501c32ec02bdd20fd3dcf3ab113d6198 (SDFP) and ec4ce3b9501c32ec02bdd20fd3dcf3ab113d6198
fix comments from review

erwango commented on dts/bindings/flash_controller/st,stm32-ospi-nor.yaml:
```diff
@@ -0,0 +1,57 @@
+      type: int
+      default: 8
+      required: false
+      description: The width of (Octo)SPI bus to which flash memory is connected.
+                   Now only value of 8 (default) or 1 (not really useful) are supported.
+      enum:
+        - 1 #  SPI mode
+        - 8 #  Octo SPI mode
```
Comments don't show up in generated bindings documentation (for instance: https://docs.zephyrproject.org/3.0.0/reference/devicetree/bindings/clock/st%2Cstm32h7-rcc.html#dtbinding-st-stm32h7-rcc)
So, I'd suggest:
```suggestion
      description: The width of (Octo)SPI bus to which flash memory is connected.
                   Octo SPI (8, default) and  SPI (1)  are supported.
      enum:
        - 1 #  SPI mode
        - 8 #  Octo SPI mode
```

erwango commented on dts/bindings/flash_controller/st,stm32-ospi-nor.yaml:
```diff
@@ -0,0 +1,57 @@
+    data-rate:
+      type: int
+      required: true
+      description: the SPI data Rate is STR or DTR (DTR is valid for OSPI only)
+      enum:
+        - 1 #  Single Transfer Rate
+        - 2 #  Double Transfer Rate
```
Comments don't show up in generated bindings documentation (for instance: https://docs.zephyrproject.org/3.0.0/reference/devicetree/bindings/clock/st%2Cstm32h7-rcc.html#dtbinding-st-stm32h7-rcc)
So, I'd suggest:
```suggestion
    data-rate:
      type: int
      required: true
      description: SPI data rate. Either Single Transfer Rate (1) or Dual Transfer Rate (8).
      DTR is only possible in OctoSPI mode.
      enum:
        - 1 #  Single Transfer Rate
        - 2 #  Double Transfer Rate
```

erwango commented on boards/arm/stm32l562e_dk/stm32l562e_dk_common.dtsi:
```diff
@@ -127,6 +127,47 @@
+
+			   partition@0 {
+			       label = "nor";
+			       reg = <0x00000000 0x00400000>;
```
```suggestion
			       reg = <0x00000000 DT_SIZE_M(4)>;
```

erwango commented on boards/arm/stm32l562e_dk/stm32l562e_dk_common.dtsi:
```diff
@@ -127,6 +127,47 @@
+		label = "MX25LM51245";
+		reg = <0>;
+		ospi-max-frequency = <50000000>;
+		size = <DT_SIZE_M(4)>;  /* 512 KBytes */
```
```suggestion
		size = <DT_SIZE_M(4)>; 
```

erwango commented on boards/arm/stm32l562e_dk/stm32l562e_dk_common.dtsi:
```diff
@@ -127,6 +127,47 @@
+		compatible = "st,stm32-ospi-nor";
+		label = "MX25LM51245";
+		reg = <0>;
+		ospi-max-frequency = <50000000>;
```
```suggestion
		ospi-max-frequency = <DT_FREQ_M(50)>;
```

erwango commented on boards/arm/b_u585i_iot02a/b_u585i_iot02a-common.dtsi:
```diff
@@ -101,6 +101,47 @@
+		compatible = "st,stm32-ospi-nor";
+		label = "MX25LM51245";
+		reg = <0>;
+		ospi-max-frequency = <50000000>;
+		size = <DT_SIZE_M(4)>;  /* 512 KBytes */
```
```suggestion
		ospi-max-frequency = <DT_SIZE_M(50)>;
		size = <DT_SIZE_M(4)>;
```

erwango commented on boards/arm/b_u585i_iot02a/b_u585i_iot02a-common.dtsi:
```diff
@@ -101,6 +101,47 @@
+
+			   partition@0 {
+			       label = "nor";
+			       reg = <0x00000000 0x00400000>;
```
```suggestion
			       reg = <0x00000000 DT_SIZE_M(4)>;
```

erwango commented on samples/drivers/spi_flash/src/main.c:
```diff
@@ -36,6 +39,10 @@
 #define FLASH_TEST_REGION_OFFSET 0x7F000
+#elif defined(CONFIG_BOARD_STM32L562E_DK) || \
+	defined(CONFIG_BOARD_B_U585I_IOT02A)
+/* NOR flash size is 0x80000 */
```
According to dts, it is 4M

FRASTM replied:
64MBytes indeed, updating

erwango commented on drivers/flash/Kconfig.stm32_ospi:
```diff
@@ -0,0 +1,20 @@
+	depends on SOC_FAMILY_STM32
+	default $(dt_compat_enabled,$(DT_COMPAT_ST_STM32_OSPI_NOR))
+	select USE_STM32_HAL_OSPI
+	select USE_STM32_LL_DELAYBLOCK
+	select USE_STM32_LL_DLYB
```
```
	select USE_STM32_LL_DELAYBLOCK if SOC_SERIES_STM32L5X
	select USE_STM32_LL_DLYB if SOC_SERIES_STM32U5X
```

erwango commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1663 @@
+	}
+
+	/* Wait that the configuration is effective and check that memory is ready */
+	HAL_Delay(STM32_OSPI_WRITE_REG_MAX_TIME);
```
```suggestion
	k_msleep(STM32_OSPI_WRITE_REG_MAX_TIME);
```

erwango commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1663 @@
+
+#endif
+	/* After SWreset CMD, wait in case SWReset occurred during erase operation */
+	HAL_Delay(STM32_OSPI_RESET_MAX_TIME);
```
```suggestion
	k_msleep(STM32_OSPI_RESET_MAX_TIME);
```

erwango commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1663 @@
+			size -= dev_cfg->flash_size;
+		} else {
+			/* sector erase */
+			LOG_INF("Sector Erase");			cmd_erase.Address = addr;
```
```suggestion
			LOG_INF("Sector Erase");
			cmd_erase.Address = addr;
```

FRASTM committed:
drivers: flash: ospi factorized for stm32 devices

Simplifies the driver and Gives a generic function to prepare the Regular
commands for each instruction.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
adding a commit to optimize the Regular Command setting before each command sending operation.

erwango commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1505 @@
+	HAL_OSPI_IRQHandler(&dev_data->hospi);
+}
+
+/* weak function requires for HAL compilation */
```
```suggestion
/* weak function required for HAL compilation */
```

erwango commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1505 @@
+	return HAL_OK;
+}
+
+/* weak function requires for HAL compilation */
```
```suggestion
/* weak function required for HAL compilation */
```

erwango commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1505 @@
+		}
+	}
+
+	uint32_t erase_size = BIT(value);
```
Variables are generally declared at the start of the block.

FRASTM replied:
this function `setup_pages_layout()` is directly coming from the drivers/flash_controller/flash_stm32_qspi.c

erwango replied:
ok, ideally it should be fixed as well.

GeorgeCGV replied:
Can we create a shared module for the common functions of QSPI and OSPI?

erwango commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1505 @@
+		return -EIO;
+	}
+
+	if (clock_control_get_rate(DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE),
+			(clock_control_subsys_t) &dev_cfg->pclken,
+			&ahb_clock_freq) < 0) {
+		LOG_DBG("Failed to get AHB clock frequency");
+		return -EIO;
```
Note that this is correct only if SYSCLK rate equals AHB bus clock rate.

GeorgeCGV replied:
> Note that this is correct only if SYSCLK rate equals AHB bus clock rate.

Could that be written as a comment?

erwango commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1505 @@
+	/* OCTOSPI I/O manager init Function */
+	OSPIM_CfgTypeDef ospi_mgr_cfg = {0};
+
+	__HAL_RCC_OSPIM_CLK_ENABLE();
```
I'm surprise this is required. This should be done one `clock_control_on()` call.

FRASTM replied:
The OSPIManager is not declared as a node of the stm32U5,

erwango replied:
ok, thanks. There may be alternat wy to handle this via device tree, but that's ok for now.

erwango commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1505 @@
+		ospi_mgr_cfg.NCSPort = 1;
+		ospi_mgr_cfg.IOLowPort = HAL_OSPIM_IOPORT_1_LOW;
+		ospi_mgr_cfg.IOHighPort = HAL_OSPIM_IOPORT_1_HIGH;
+
```
Not sure about the blank line here.

FRASTM replied:
line deleted

erwango commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1505 @@
+		ospi_mgr_cfg.NCSPort = 2;
+		ospi_mgr_cfg.IOLowPort = HAL_OSPIM_IOPORT_2_LOW;
+		ospi_mgr_cfg.IOHighPort = HAL_OSPIM_IOPORT_2_HIGH;
+
```
Not sure about the blank line here.

FRASTM replied:
line deleted

FRASTM commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1499 @@
+	s_command.Instruction = (nor_mode == OSPI_SPI_MODE)
+				? SPI_NOR_CMD_RDSR
+				: SPI_NOR_OCMD_RDSR;
+	s_command.AddressMode = (nor_mode == OSPI_SPI_MODE)
+				? HAL_OSPI_ADDRESS_NONE
+				: HAL_OSPI_ADDRESS_8_LINES;
```
not exactly, in the ospi_prepare_cmd the the address is set to `HAL_OSPI_ADDRESS_1_LINE` or `HAL_OSPI_ADDRESS_8_LINES`. 
Here in the stm32_ospi_mem_ready(), the AddressMode is set to HAL_OSPI_ADDRESS_NONE. That's the way STM32Cube does. Changing to HAL_OSPI_ADDRESS_1_LINE instead of HAL_OSPI_ADDRESS_NONE works tough
So removing this s_command.AddressMode

erwango replied:
Right, I read too fast, txs for the update.

FRASTM commented on dts/bindings/flash_controller/st,stm32-ospi-nor.yaml:
```diff
@@ -0,0 +1,60 @@
+    spi-bus-width:
+      type: int
+      required: true
+      description: The width of (Octo)SPI bus to which flash memory is connected.
+                   Possible values are in the dt-bindings/flash_controller/ospi.h 
+                   <OSPI_SPI_MODE> or <OSPI_OPI_MODE> are supported.
+      enum:
+        - 1 # OSPI_OPI_MODE = SPI mode on 1 data line
+        - 8 # OSPI_SPI_MODE = Octo SPI mode on 8 data lines
```
with 
```
enum:
        - OSPI_SPI_MODE
        - OSPI_OPI_MODE
```
the #define OSPI_SPI_MODE is not known by the st,stm32-ospi-nor.yaml

FRASTM replied:
with property "int", the DTS can still declare it with the  `<OSPI_OPI_MODE>` or `<OSPI_SPI_MODE>` but the enum in this .yaml file must remain numerical  
```
      enum:
        - 1
        - 8
```

erwango replied:
Ok, but then we need to fix the description which is not clear enough (as people can't see the comment on line):
The following would be clearer:
```suggestion
      description: The width of (Octo)SPI bus to which flash memory is connected.
                   Possible values are :
                   - OSPI_SPI_MODE <1> = SPI mode on 1 data line
                   - OSPI_OPI_MODE <8> = Octo SPI mode on 8 data lines
```

erwango commented on drivers/flash/Kconfig.stm32_ospi:
```diff
@@ -0,0 +1,20 @@
+	default $(dt_compat_enabled,$(DT_COMPAT_ST_STM32_OSPI_NOR))
+	select USE_STM32_HAL_OSPI
+	select USE_STM32_LL_DELAYBLOCK if (SOC_SERIES_STM32MP1X || SOC_SERIES_STM32H7X)
+	select USE_STM32_LL_DLYB
```
but so far, this drivers doesn't support H7 nor MP1. This line could create confusion.
Besides, we're not sure these series would have the same dependency (as I understand L5 has no dependency).

FRASTM replied:
done

FRASTM commented on boards/arm/b_u585i_iot02a/b_u585i_iot02a-common.dtsi:
```diff
@@ -7,6 +7,7 @@
 #include <st/u5/stm32u585Xi.dtsi>
 #include <st/u5/stm32u585aiixq-pinctrl.dtsi>
 #include "arduino_r3_connector.dtsi"
+#include "dt-bindings/flash_controller/ospi.h"
```
moved to dts/arm/st/u5/stm32u5.dtsi

FRASTM commented on boards/arm/stm32l562e_dk/stm32l562e_dk_common.dtsi:
```diff
@@ -7,6 +7,7 @@
 #include <st/l5/stm32l562Xe.dtsi>
 #include <st/l5/stm32l562qeixq-pinctrl.dtsi>
 #include "arduino_r3_connector.dtsi"
+#include "dt-bindings/flash_controller/ospi.h"
```
moved to dts/arm/st/l5/stm32l5.dtsi

FRASTM commented on drivers/flash/Kconfig.stm32_ospi:
```diff
@@ -0,0 +1,20 @@
+	depends on SOC_FAMILY_STM32
+	default $(dt_compat_enabled,$(DT_COMPAT_ST_STM32_OSPI_NOR))
+	select USE_STM32_HAL_OSPI
+	select USE_STM32_LL_DELAYBLOCK if (SOC_SERIES_STM32MP1X || SOC_SERIES_STM32H7X)
```
removed : if...

erwango replied:
From what I understand, on L5 there is no dependency on USE_STM32_LL_DELAYBLOCK, am I right ?
In that case, the whole line should be removed.

erwango commented on dts/bindings/flash_controller/st,stm32-ospi-nor.yaml:
```diff
@@ -0,0 +1,60 @@
+                data-mode = <OSPI_OPI_MODE>; /* access on 8 data lines */
+                data-rate = <OSPI_DTR_TRANSFER>; /* access in DTR */
+                ospi-max-frequency = <50000000>;
+                size = <DT_SIZE_M(4)>; /* 512 KBytes */
```
```suggestion
                size = <DT_SIZE_M(4)>;
```

erwango commented on dts/bindings/flash_controller/st,stm32-ospi-nor.yaml:
```diff
@@ -0,0 +1,60 @@
+                reg = <0>;
+                data-mode = <OSPI_OPI_MODE>; /* access on 8 data lines */
+                data-rate = <OSPI_DTR_TRANSFER>; /* access in DTR */
+                ospi-max-frequency = <50000000>;
```
```suggestion
                ospi-max-frequency = <DT_FREQ_M(50)>;
```

erwango commented on dts/bindings/flash_controller/st,stm32-ospi-nor.yaml:
```diff
@@ -0,0 +1,60 @@
+    data-rate:
+      type: int
+      required: true
+      description: The SPI data Rate is STR or DTR (DTR is valid for OSPI only)
+                   Possible values are in the dt-bindings/flash_controller/ospi.h
+                   <OSPI_STR_TRANSFER> or <OSPI_DTR_TRANSFER>
```
Same here:
```suggestion
      description: The SPI data Rate is STR or DTR
                   Possible values are :
                   - OSPI_STR_TRANSFER <1> = Single Rate Transfer
                   - OSPI_DTR_TRANSFER <2> = Dual Rate Transfer (valid only with OSPI_OPI_MODE)
```

erwango commented on dts/bindings/flash_controller/st,stm32-ospi-nor.yaml:
```diff
@@ -0,0 +1,60 @@
+                   Possible values are in the dt-bindings/flash_controller/ospi.h
+                   <OSPI_STR_TRANSFER> or <OSPI_DTR_TRANSFER>
+      enum:
+        - 1 # OSPI_STR_TRANSFER = Single-data Transfer Rate
+        - 2 # OSPI_DTR_TRANSFER = Double-data Transfer Rate
```
Remove comments

erwango commented on dts/bindings/flash_controller/st,stm32-ospi-nor.yaml:
```diff
@@ -0,0 +1,60 @@
+                   Possible values are in the dt-bindings/flash_controller/ospi.h 
+                   <OSPI_SPI_MODE> or <OSPI_OPI_MODE> are supported.
+      enum:
+        - 1 # OSPI_OPI_MODE = SPI mode on 1 data line
+        - 8 # OSPI_SPI_MODE = Octo SPI mode on 8 data lines
```
Remove comments

erwango commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1676 @@
+LOG_MODULE_REGISTER(flash_stm32_ospi, CONFIG_FLASH_LOG_LEVEL);
+
+#define STM32_OSPI_RESET_GPIO DT_INST_NODE_HAS_PROP(0, reset_gpios)
+#if STM32_OSPI_RESET_GPIO
+#include <drivers/gpio.h>
+#endif /* STM32_OSPI_RESET_GPIO */
```
Let's remove the conditional here and include in any case.
This is simpler and  how it is done in general.

FRASTM replied:
done

erwango commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1500 @@
+
+#define DT_DRV_COMPAT st_stm32_ospi_nor
+
+#include <errno.h>
+#include <kernel.h>
+#include <toolchain.h>
+#include <arch/common/ffs.h>
+#include <sys/util.h>
+#include <soc.h>
+#include <drivers/pinctrl.h>
+#include <drivers/clock_control/stm32_clock_control.h>
+#include <drivers/clock_control.h>
+#include <drivers/flash.h>
+#include <dt-bindings/flash_controller/ospi.h>
+
+#include "spi_nor.h"
+#include "jesd216.h"
+
+#include <logging/log.h>
```
```suggestion
#include <errno.h>
#include <zephyr/kernel.h>
#include <toolchain.h>
#include <zephyr/arch/common/ffs.h>
#include <zephyr/sys/util.h>
#include <soc.h>
#include <zephyr/drivers/pinctrl.h>
#include <zephyr/drivers/clock_control/stm32_clock_control.h>
#include <zephyr/drivers/clock_control.h>
#include <zephyr/drivers/flash.h>
#include <zephyr/dt-bindings/flash_controller/ospi.h>

#include "spi_nor.h"
#include "jesd216.h"

#include <zephyr/logging/log.h>
```
Cf https://github.com/zephyrproject-rtos/zephyr/pull/45410

FRASTM replied:
changed : include path

FRASTM commented on dts/arm/st/u5/stm32u5.dtsi:
```diff
@@ -10,6 +10,7 @@
 #include <dt-bindings/clock/stm32u5_clock.h>
 #include <dt-bindings/gpio/gpio.h>
 #include <dt-bindings/i2c/i2c.h>
+#include "dt-bindings/flash_controller/ospi.h"
```
done

FRASTM commented:
change includes path according to drivers: migrate includes to <zephyr/...> #45410

GeorgeCGV commented on drivers/flash/spi_nor.h:
```diff
@@ -42,11 +42,57 @@
+#define SPI_NOR_CMD_RESET_EN    0x66    /* Reset Enable */
+#define SPI_NOR_CMD_RESET_MEM   0x99    /* Reset Memory */
+#define SPI_NOR_CMD_BULKE       0x60    /* Bulk Erase */
+#define SPI_NOR_CMD_4B_PP       0x12    /* Page Program 4 Byte Address */
```
Maybe it is better to stick with definitions similar to Linux?

```suggestion
#define SPI_NOR_CMD_PP_4B       0x12    /* Page Program 4 Byte Address */
```

GeorgeCGV commented on drivers/flash/spi_nor.h:
```diff
@@ -42,11 +42,57 @@
+#define SPI_NOR_CMD_RESET_MEM   0x99    /* Reset Memory */
+#define SPI_NOR_CMD_BULKE       0x60    /* Bulk Erase */
+#define SPI_NOR_CMD_4B_PP       0x12    /* Page Program 4 Byte Address */
+#define SPI_NOR_CMD_4B_FAST_RD  0x0C    /* Fast Read 4 Byte Address */
```
```suggestion
#define SPI_NOR_CMD_READ_FAST_4B  0x0C    /* Fast Read 4 Byte Address */
```

FRASTM replied:
renamed to `SPI_NOR_CMD_READ_FAST_4B`

GeorgeCGV commented on dts/bindings/flash_controller/st,stm32-ospi-nor.yaml:
```diff
@@ -0,0 +1,62 @@
+    spi-bus-width:
+      type: int
+      required: true
+      description: The width of (Octo)SPI bus to which flash memory is connected.
+                   Possible values are :
+                   - OSPI_SPI_MODE <1> = SPI mode on 1 data line
+                   - OSPI_OPI_MODE <8> = Octo SPI mode on 8 data lines
```
Wasn't the rule to use `description: |`?

```suggestion
      description: |
        The width of (Octo)SPI bus to which flash memory is connected.
        
        Possible values are :
          - OSPI_SPI_MODE <1> = SPI mode on 1 data line
          - OSPI_OPI_MODE <8> = Octo SPI mode on 8 data lines
```

FRASTM replied:
modified

GeorgeCGV commented on dts/bindings/flash_controller/st,stm32-ospi-nor.yaml:
```diff
@@ -0,0 +1,62 @@
+    data-rate:
+      type: int
+      required: true
+      description: The SPI data Rate is STR or DTR
+                   Possible values are :
+                   - OSPI_STR_TRANSFER <1> = Single Rate Transfer
+                   - OSPI_DTR_TRANSFER <2> = Dual Rate Transfer (only with OSPI_OPI_MODE)
```
Wasn't the rule to use `description: |`?

```suggestion
      description: |
        The SPI data Rate is STR or DTR.
        
        Possible values are :
          - OSPI_STR_TRANSFER <1> = Single Rate Transfer
          - OSPI_DTR_TRANSFER <2> = Dual Rate Transfer (only with OSPI_OPI_MODE)
```

GeorgeCGV commented on modules/Kconfig.stm32:
```diff
@@ -550,6 +550,11 @@ config USE_STM32_LL_DAC
+config USE_STM32_LL_DLYB
 	bool
 	help
 	  Enable STM32Cube DelayBlock (DELAYBLOCK) LL module driver
```
The `USE_STM32_LL_DELAYBLOCK` help contains the series name, maybe it is worth adding it to `USE_STM32_LL_DLYB` as well?

```suggestion
	  Enable STM32Cube DelayBlock (DELAYBLOCK) LL module driver (stm32U5)
```

FRASTM replied:
modified

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+	k_sem_give(&dev_data->sem);
+}
+
+/*
+ * Send a command over OSPI bus.
+ */
```
Could we omit comments that are too obvious or give no context?
```suggestion
```

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+	return dev_data->cmd_status;
+}
+
+/*
+ * Perform a read access over OSPI bus.
+ */
```
Same.
```suggestion
```

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+	return dev_data->cmd_status;
+}
+
+/*
+ * Perform a write access over OSPI bus.
+ */
```
Same.
```suggestion
```

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+/*
+ * Read Serial Flash ID Parameter
+ */
+static int ospi_read_norid(const struct device *dev, uint8_t *data,
```
Isn't it commonly known as `jedec_id`? Maybe better to use that instead of `norid`?

FRASTM replied:
all this part about` ospi_read_norid()  ` is removed

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+	HAL_StatusTypeDef hal_ret;
+
+	OSPI_RegularCmdTypeDef cmd = {
+		.Instruction = 0x9F,
```
Could use `spi_nor.h` `SPI_NOR_CMD_RDID` define.
```suggestion
		.Instruction = SPI_NOR_CMD_RDID,
```

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+	};
+	dev_data->cmd_status = 0;
+
+	/* do not use command IT here */
```
> /* do not use receive IT here */

No help. Why not?

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+		LOG_ERR("%d: Failed to send OSPI instruction", hal_ret);
+		return -EIO;
+	}
+	/* do not use receive IT here */
```
Same.

FRASTM replied:
removing comment

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+	int cmd_status;
+};
+
+#define DEV_NAME(dev) ((dev)->name)
+#define DEV_CFG(dev) \
+	(const struct flash_stm32_ospi_config * const)(dev->config)
+#define DEV_DATA(dev) \
+	(struct flash_stm32_ospi_data * const)(dev->data)
```
Most drivers are getting rid of that.

For example, recent QSPI PR #43824.

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+				: HAL_OSPI_DATA_8_LINES;
+	s_command.NbData         = (nor_rate == OSPI_DTR_TRANSFER) ? 2U : 1U;
+
+	/* Send the command */
```
Obvious comment with no purpose.

```suggestion
```

FRASTM replied:
removing comment

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+		LOG_ERR("OSPI transmit ");
+		return -EIO;
+	}
+	uint8_t tmp = SPI_NOR_CR2_DUMMY_CYCLES_66MHZ;
```
Could that be moved to the top of the function and named in a way to add some meaning to it (i.e. `dummy_cycles`)?

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+	s_command.NbData = (nor_mode == OSPI_SPI_MODE) ? 1U
+			: ((nor_rate == OSPI_DTR_TRANSFER) ? 2U : 1U);
+
+	/* Send the command */
```
```suggestion
```

FRASTM replied:
removing comment

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+		LOG_ERR("Write Flash configuration reg2 failed");
+		return -EIO;
+	}
+	uint8_t tmp = op_enable;
```
> uint8_t tmp = op_enable;

Could that be moved to the top of the function and named in a way to add some meaning?

FRASTM replied:
`HAL_OSPI_Transmit(hospi, &op_enable` is even better

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+					: SPI_NOR_DUMMY_REG_OCTAL);
+	s_command.NbData = (nor_rate == OSPI_DTR_TRANSFER) ? 2U : 1U;
+
+	/* Send the command */
```
```suggestion
```

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+		return -EIO;
+	}
+
+	/* Reception of the data */
```
```suggestion
```

FRASTM replied:
removing comment

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+	}
+
+	/* going to set the OPI mode (STR or DTR transfer rate) */
+	LOG_INF("OSPI configuring OctoSPI mode");
```
```suggestion
	LOG_DBG("OSPI configuring OctoSPI mode");
```

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+	struct flash_stm32_ospi_data *dev_data = DEV_DATA(dev);
+	int ret = 0;
+
+	/* erase non-zero size */
```
```suggestion
	/* ignore zero size erase */
```

FRASTM replied:
changed to LOG_DBG

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+
+	/* Maximise erase size : means the complete chip */
+	if (size > dev_cfg->flash_size) {
+		/* reset the addr in that case */
```
```suggestion
		/* reset addr in that case */
```

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+
+	ospi_lock_thread(dev);
+
+	/* Check Flash busy */
```
```suggestion
```

FRASTM replied:
removing comment

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+	if (stm32_ospi_mem_ready(&dev_data->hospi,
+		dev_cfg->data_mode, dev_cfg->data_rate) != 0) {
+		LOG_DBG("Erase failed : flash busy");
+		return -ENOTSUP;
```
Shouldn't it be `-EAGAIN` if the Flash is busy?

FRASTM replied:
or `-EBUSY` ?

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+		return -ENOTSUP;
+	}
+
+	/* Enable write operations */
```
```suggestion
```

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+	if (stm32_ospi_write_enable(&dev_data->hospi,
+		dev_cfg->data_mode, dev_cfg->data_rate) != 0) {
+		LOG_DBG("Erase failed : write enable");
+		return -ENOTSUP;
```
> return -ENOTSUP;

Shouldn't it be `-EIO` when write enable fails? Even better is to propagate the `stm32_ospi_write_enable` return code.

FRASTM replied:
yes, it is  `-EIO` or `0`

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+					cmd_erase.Address = addr;
+				}
+			}
+			ospi_send_cmd(dev, &cmd_erase);
```
```suggestion

			ospi_send_cmd(dev, &cmd_erase);
			
```

FRASTM replied:
fixed

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+			} else {
+				addr += SPI_NOR_SECTOR_SIZE;
+				size -= SPI_NOR_SECTOR_SIZE;
+			}
+		ret = stm32_ospi_mem_ready(&dev_data->hospi,
+						 dev_cfg->data_mode, dev_cfg->data_rate);
+		}
```
```suggestion
			}
			ret = stm32_ospi_mem_ready(&dev_data->hospi,
						   dev_cfg->data_mode, dev_cfg->data_rate);
		}
```

FRASTM replied:
fixed

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+		return -EINVAL;
+	}
+
+	/* read non-zero size */
```
```suggestion
		/* reset addr in that case */
```

or even remove the comment as it is quite obvious what the check is for...

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+	cmd.Address = addr;
+	/* DataSize is set by the read cmd */
+
+	LOG_INF("OSPI: read %u data", size);
```
```suggestion
	LOG_DBG("OSPI: read %u data", size);
```

FRASTM replied:
changed to LOG_DBG

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+		return -EINVAL;
+	}
+
+	/* write non-zero size */
```
```suggestion
	/* ignore zero size erase */
```
or even remove the comment as it is quite obvious what the check is for...

FRASTM replied:
changing comment

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+	int ret = 0;
+
+	if (!ospi_address_is_valid(dev, addr, size)) {
+		LOG_DBG("Error: address or size exceeds expected values: "
```
```suggestion
		LOG_ERR("Error: address or size exceeds expected values: "
```

That looks important to be seen.

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+	int ret;
+
+	if (!ospi_address_is_valid(dev, addr, size)) {
+		LOG_DBG("Error: address or size exceeds expected values: "
```
```suggestion
		LOG_ERR("Error: address or size exceeds expected values: "
```
That looks important to be seen.

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+	/* Enable write operations */
+	if (stm32_ospi_write_enable(&dev_data->hospi,
+		dev_cfg->data_mode, dev_cfg->data_rate) != 0) {
+		LOG_DBG("Erase failed : write enable");
```
```suggestion
		LOG_ERR("Erase failed : write enable");
```
That looks important to be seen.

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+	/* Check Flash busy */
+	if (stm32_ospi_mem_ready(&dev_data->hospi,
+		dev_cfg->data_mode, dev_cfg->data_rate) != 0) {
+		LOG_DBG("Erase failed : flash busy");
```
```suggestion
		LOG_ERR("Erase failed : flash busy");
```
That looks important to be seen.

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+	struct flash_stm32_ospi_data *dev_data =
+		CONTAINER_OF(hospi, struct flash_stm32_ospi_data, hospi);
+
+	LOG_DBG("Error cb");
```
> LOG_DBG("Error cb");

Can we clarify the error a bit more? If not, then maybe it is not worth having the log output here at all?

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+		ospi_mgr_cfg.IOHighPort = HAL_OSPIM_IOPORT_2_HIGH;
+	}
+
+	ospi_mgr_cfg.Req2AckTime = 1; /* arbitrary value */
```
```suggestion
	ospi_mgr_cfg.Req2AckTime = 1;
```

FRASTM replied:
removing comment

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+	}
+
+	/* send the instruction to read the NOR-flash ID : check access to mem. */
+	uint8_t nor_id[3];
+
+	ret = ospi_read_norid(dev, nor_id, sizeof(nor_id));
+	if (ret != 0) {
+		LOG_ERR("reading Nor-flash ID failed: %d", ret);
+		return ret;
+	}
+
+	LOG_DBG("Nor-flash ID : Manuf 0x%X, Mem Type 0x%X, Density 0x%X",
+		nor_id[2], nor_id[0], nor_id[1]);
```
Please, could the `jedec_id` be used instead of `nor_id`?

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+	}
+
+	/* send the instruction to read the NOR-flash ID : check access to mem. */
+	uint8_t nor_id[3];
+
+	ret = ospi_read_norid(dev, nor_id, sizeof(nor_id));
+	if (ret != 0) {
+		LOG_ERR("reading Nor-flash ID failed: %d", ret);
+		return ret;
+	}
+
+	LOG_DBG("Nor-flash ID : Manuf 0x%X, Mem Type 0x%X, Density 0x%X",
+		nor_id[2], nor_id[0], nor_id[1]);
```
Please, could the `jedec_id` be used instead of `nor_id`?

```suggestion
	uint8_t jedec_id[3];

	ret = ospi_read_norid(dev, jedec_id, sizeof(jedec_id));
	if (ret != 0) {
		LOG_ERR("JEDEC ID read failed: %d", ret);
		return ret;
	}

	LOG_DBG("JEDEC ID : Manuf 0x%X, Mem Type 0x%X, Density 0x%X",
		nor_id[2], nor_id[0], nor_id[1]);
```

FRASTM replied:
Ok, I will remove this function entirely. It was useful to control that a NOR flash  not responding to SFDP command, was even capable (alive) to answer to NOR id command.

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+		LOG_ERR("SFDP magic %08x invalid", magic);
+		return -EINVAL;
+	}
+
+
```
```suggestion

```

FRASTM commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1498 @@
+	/* Run IRQ init */
+	dev_cfg->irq_config(dev);
+
+	LOG_INF("Device %s initialized", DEV_NAME(dev));
```
removing

FRASTM commented:
Deleting the `ospi_read_norid() `
Changes after review

erwango commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1422 @@
+	}
+#else /* sfdp_bfp */
+
+	OSPI_RegularCmdTypeDef cmd = {
```
Why not calling `ospi_prepare_cmd()` here ?

FRASTM replied:
yes, including the ospi_prepare_cmd here

erwango commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1422 @@
+static int stm32_ospi_write_cfg2reg_io(OSPI_HandleTypeDef *hospi,
+				       uint8_t nor_mode, uint8_t nor_rate, uint8_t op_enable)
+{
+	OSPI_RegularCmdTypeDef s_command = s_command = ospi_prepare_cmd(nor_mode, nor_rate);
```
```suggestion
	OSPI_RegularCmdTypeDef s_command = ospi_prepare_cmd(nor_mode, nor_rate);
```

erwango commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1422 @@
+static int stm32_ospi_read_cfg2reg(OSPI_HandleTypeDef *hospi,
+				   uint8_t nor_mode, uint8_t nor_rate, uint8_t *value)
+{
+	OSPI_RegularCmdTypeDef s_command = s_command = ospi_prepare_cmd(nor_mode, nor_rate);
```
```suggestion
	OSPI_RegularCmdTypeDef s_command = ospi_prepare_cmd(nor_mode, nor_rate);
```

erwango commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1422 @@
+					uint8_t nor_mode, uint8_t nor_rate)
+{
+	uint8_t transmit_data = SPI_NOR_CR2_DUMMY_CYCLES_66MHZ;
+	OSPI_RegularCmdTypeDef s_command = s_command = ospi_prepare_cmd(nor_mode, nor_rate);
```
```suggestion
	OSPI_RegularCmdTypeDef s_command = ospi_prepare_cmd(nor_mode, nor_rate);
```

erwango commented on dts/bindings/ospi/st,stm32-ospi.yaml:
```diff
@@ -0,0 +1,35 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+    STM32 OSPI device representation. A stm32 octospi node would typically
+    looks to this:
```
```suggestion
    STM32 OSPI device representation. Enableing a stm32 octospi node in a board 
    description would typically require this:
```

FRASTM replied:
ok, description changed

FRASTM commented:
rebase on 48877f8bdcd92f7e14a72ab4c830989f15f1e884 
(where the hal_stm32 is updated --> removing commit 398a1a0f97174ef954d2e45a3faa5f919c71d2d1

FRASTM commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -189,42 +231,13 @@ static int ospi_read_sfdp(const struct device *dev, off_t addr, uint8_t *data,
+	OSPI_RegularCmdTypeDef cmd = ospi_prepare_cmd(dev_cfg->data_mode,
+						      dev_cfg->data_rate);
+	cmd.Address = addr;
+	cmd.AddressSize = ((dev_cfg->data_mode == OSPI_SPI_MODE)
```
correct, restoring
Indeed, the sdfp was not read by command by table (DTS) with this mcus

FRASTM commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -169,6 +169,48 @@ static int ospi_write_access(const struct device *dev, OSPI_RegularCmdTypeDef *c
 
+/*
+ * Gives a OSPI_RegularCmdTypeDef with all parameters set
+ * except Intruction, Address, DummyCycles, NbData
```
fixed

FRASTM commented:
restore instruction in ospi_read_sfdp

ABOSTM reviewed (approved):
LGTM

GeorgeCGV commented on drivers/flash/spi_nor.h:
```diff
@@ -42,11 +42,57 @@
+#define SPI_NOR_CMD_RESET_MEM   0x99    /* Reset Memory */
+#define SPI_NOR_CMD_BULKE       0x60    /* Bulk Erase */
+#define SPI_NOR_CMD_PP_4B       0x12    /* Page Program 4 Byte Address */
+#define SPI_NOR_CMD_READ_FAST_4B  0x0C  /* Fast Read 4 Byte Address */
```
LGTM

Not a showstopper, but would be great to align the define values under `/* Flash opcodes */`.

FRASTM replied:
will be updated in the next PR

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1398 @@
+ * Gives a OSPI_RegularCmdTypeDef with all parameters set
+ * except Instruction, Address, DummyCycles, NbData
+ */
+static OSPI_RegularCmdTypeDef ospi_prepare_cmd(uint8_t transfer_mode, uint8_t transfer_rate)
```
Minor, but for performance reasons the struct copying could be avoided by using a pointer to the command struct. The function could fill in the values using it.

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -0,0 +1,1398 @@
+	s_command.Instruction = (nor_mode == OSPI_SPI_MODE)
+				? SPI_NOR_CMD_RDSR
+				: SPI_NOR_OCMD_RDSR;
+	s_command.Address = 0;
```
Don't we need  `s_command.AddressMode = HAL_OSPI_ADDRESS_NONE;`?
